flocker/node/_loop.py

```diff
@@ -0,0 +1,282 @@
+
+    def output_DISCONNECT(self, symbol, context):
+        self.client.transport.abortConnection()
+
+
+def build_cluster_status_fsm(agent_operation_fsm):
+    S = ClusterStatusStates
+    I = ClusterStatusInputs
+    O = ClusterStatusOutputs
+    table = TransitionTable()
+    table = table.addTransitions(
+        S.DISCONNECTED, {
+            I.CLIENT_CONNECTED: ([O.STORE_CLIENT], S.IGNORANT),
+            I.SHUTDOWN: ([], S.SHUTDOWN),
+        })
+    table = table.addTransitions(
+        S.IGNORANT, {
+            I.CLIENT_DISCONNECTED: ([], S.DISCONNECTED),
+            I.STATUS_UPDATE: ([O.READY], S.KNOWLEDGEABLE),
+            I.SHUTDOWN: ([O.DISCONNECT], S.SHUTDOWN),
+        })
+    table = table.addTransitions(
+        S.KNOWLEDGEABLE, {
+            I.STATUS_UPDATE: ([], S.KNOWLEDGEABLE),
+            I.CLIENT_DISCONNECTED: ([O.NOT_READY], S.DISCONNECTED),
+            I.SHUTDOWN: ([O.NOT_READY, O.DISCONNECT], S.SHUTDOWN),
+        })
+
+    return constructFiniteStateMachine(
+        inputs=I, outputs=O, states=S, table=table,
+        richInput=[_ClientConnected, _StatusUpdate],
+        inputContext={}, world=MethodSuffixOutputer(
+            ClusterStatus(agent_operation_fsm)))
+
+
+class AgentOperationInputs(Names):
+    GO = NamedConstant()
+    STOP = NamedConstant()
+    DISCOVERED_STATUS = NamedConstant()
+    CHANGES_DONE = NamedConstant()
+
+
+class _Go(trivialInput(AgentOperationInputs.GO)):
+    def __init__(self, client, configuration, state):
+        pass
+
+
+class _DiscoveredStatus(trivialInput(AgentOperationInputs.DiSCOVERED_STATUS)): 
+    def __init__(self, node_state):
+        pass
+
+
+class AgentOperationStates(Names):
+    STOPPED = NamedConstant()
+    DISCOVERING = NamedConstant()
+    DISCOVERING_STOPPING = NamedConstant()
+    CHANGING = NamedConstant()
+    CHANGING_STOPPING = NamedConstant()
+
+
+class AgentOperationOutputs(Names):
+    STORE_INFO = NamedConstant()
+    DISCOVER = NamedConstant()
+    REPORT_NODE_STATE = NamedConstant()
+    CHANGE = NamedConstant()
+
+
+class AgentOperation(object):
+    def __init__(self, deployer):
+        self.deployer = deployer
+
+    def output_STORE_INFO(self, symbol, context):
+        self.client, self.configuration, self.cluster_state = (
+            context.client, context.configuration, context.state)
+
+    def output_DISCOVER(self, symbol, context):
+        d = self.deployer.discover_node_configuration()
+        d.addCallback(
+            lambda node_state: self.input(_DiscoveredStatus(node_state)))
+        # XXX error case
+
+    def output_REPORT_NODE_STATE(self, symbol, context):
+        self.client.callRemote(NodeStateCommand, node_state=context.node_state)
+
+    def output_CHANGE(self, symbol, context):
+        # XXX need to refactor Deployer slightly so you can pass in NodeState...
+        d = self.deployer.change_node_state(self.configuration, self.cluster_state,
+                                            HOSTNAME)
```

Comment:
While thinking about how the OpenStack agent is going to work, it occurred to me that if we fix the signature of `change_node_state` to omit hostname then this FSM might be usable to drive either the P2P or OpenStack dataset agents.  This would require the P2P deployer changing to be told what host it is operating on via `__init__` instead of `change_node_state`.  That seems fairly easy though.